### PR TITLE
Change to new C++ based addon interface style

### PR DESCRIPTION
**Pull request by lrusak**
I've updated this with similar changes to vfs.rar

I've tested and it is working with https://github.com/lrusak/xbmc/tree/nfs-cleanup

review might be a little hard because I rearranged NFSFile.cpp to have the functions listed in order with the API list here, https://github.com/xbmc/xbmc/blob/master/xbmc/addons/kodi-addon-dev-kit/include/kodi/addon-instance/VFS.h#L137

I wish I had moved the functions after I committed the changes but oh well. If you want me to change it just say so.

### CODE REVIEW

**Comment by notspiff** [comment]
it's fine, your word that it still works is good enough. since you are working on it, would you mind syncing it with latest master while at it ?

**Comment by lrusak** [comment]
You mean sync the NFS code from kodi master to here? I can do that.

**Comment by notspiff** [comment]
Yeah addon was forked 3-4 years ago bound to be some changes

25. sep. 2017 18:08 skrev "Lukas Rusak" <notifications@github.com>:

**Comment by lrusak** [comment]
I pushed a fixup, have a quick look and I'll squash

File: src/NFSFile.cpp [hunk 1]
```diff
@@ -596,7 +597,7 @@ bool CNFSFile::GetDirectory(const VFSURL& url, std::vector<kodi::vfs::CDirEntry>
       P8PLATFORM::FileTimeToLocalFileTime(&fileTime, &localTime);*/
 
       kodi::vfs::CDirEntry pItem;
-      pItem.SetLabel(strdup(nfsdirent->name));
+      pItem.SetLabel(strdup(tmpDirent.name));
```
Comment (garbear) [inline src/NFSFile.cpp]: Leaks

Comment (lrusak) [reply src/NFSFile.cpp]: how?

Comment (garbear) [reply src/NFSFile.cpp]: strdup() calls malloc, passes the memory to the std::string constructor, then fails to free the memory. This line should be

```c++
pItem.SetLabel(tmpDirent.name);
```

or 

```c++
pItem.SetLabel(tmpDirent.name != nullptr ? tmpDirent.name : "");
```

Comment (lrusak) [reply src/NFSFile.cpp]: ah ok, I image I should do the same with the other strdup lines?

File: src/NFSFile.cpp [hunk 2]
```diff
@@ -673,8 +673,8 @@ bool CNFSFile::GetServerList(std::vector<kodi::vfs::CDirEntry>& items)
     std::string path(std::string("nfs://") + currentExport);
     if (path[path.size()-1] != '/')
       path += '/';
-    pItem.SetPath(strdup(path.c_str()));
-    pItem.SetLabel(strdup(currentExport.c_str()));
+    pItem.SetPath(path.c_str());
+    pItem.SetLabel(currentExport.c_str());
```
Comment (garbear) [inline src/NFSFile.cpp]: All of these .c_str() can be dropped. Also there's some more lines that had me worried for memory safety, I'll check this out later.

**Comment by lrusak** [comment]
@garbear thanks for the review. I really appreciate it. Any other hints are more than welcome

File: src/NFSFile.cpp [hunk 3]
```diff
@@ -209,8 +209,8 @@ ssize_t CNFSFile::Write(void* context, const void* lpBuf, size_t uiBufSize)
     //danger - something went wrong
     if (writtenBytes < 0)
     {
-      kodi::Log(ADDON_LOG_ERROR, "Failed to pwrite(%s) %s\n",
-                ctx->filename.c_str(), nfs_get_error(ctx->pNfsContext));
```
Comment (garbear) [inline src/NFSFile.cpp]: This .c_str() needs to stay. printf-like functions require a char* pointer and will crash if given a c++ object.

Also do NFS urls have usernames/passwords in them? Should the filename be redacted in logs?

File: CMakeLists.txt [hunk 4]
```diff
@@ -7,17 +7,22 @@ list(APPEND CMAKE_MODULE_PATH ${PROJECT_SOURCE_DIR})
...
 
 find_package(NFS REQUIRED)
-find_package(kodi REQUIRED)
+find_package(Kodi REQUIRED)
+find_package(p8-platform REQUIRED)
 find_package(kodiplatform REQUIRED)
 
-include_directories(${KODI_INCLUDE_DIR}
-                    ${KODIPLATFORM_INCLUDE_DIRS}
+include_directories(${KODI_INCLUDE_DIR}/.. # Hack way with "/..", need bigger Kodi cmake rework to match right include ways
```
Comment (wsnipex) [inline CMakeLists.txt:14]: ??

Comment (notspiff) [reply CMakeLists.txt:14]: generic meh with new add-on bindings. ask @AlwinEsch when it will be resolved and how to resolve it..

Comment (AlwinEsch) [reply CMakeLists.txt:14]: Yeahh, the stupid bad thing. The old addon system has always included the kodi folder itself. 

For a `#include "libKODI_vfs.h"` was it OK. But with the new where it then is `#include "Filesystem.h"` only can it be complicated with other headers from somewhere else.

Thats why new ones should always use a `#include <kodi/Filesystem.h>`.

But as long there still addon systems (PVR and Game) is this bad thing needed to have right folder for here and on the old ones.

After them are replaced becomes the cmake part changed.

File: src/NFSConnection.cpp [hunk 5]
```diff
@@ -98,19 +98,23 @@ std::list<std::string> CNFSConnection::GetExportList()
...
+  m_readChunkSize = 0;
   m_pNfsContext = NULL;
-  m_KeepAliveTimeouts.clear();
 }
 
 void CNFSConnection::destroyOpenContexts()
 {
   m_openContextLock.Lock();
-  for(tOpenContextMap::iterator it = m_openContextMap.begin();it!=m_openContextMap.end();it++)
+  for(tOpenContextMap::iterator it = m_openContextMap.begin();it!=m_openContextMap.end();++it)
```
Comment (AlwinEsch) [inline src/NFSConnection.cpp]: Maybe change direct to the C++11 style with a:
```cpp
  for (const auto& entry : m_openContextMap)
  {
    nfs_destroy_context(entry.second.pContext);
  }
```

File: src/NFSConnection.cpp [hunk 6]
```diff
@@ -121,8 +125,8 @@ void CNFSConnection::destroyOpenContexts()
 void CNFSConnection::destroyContext(const std::string& exportName)
 {
   m_openContextLock.Lock();
-  tOpenContextMap::iterator it = m_openContextMap.find(exportName.c_str());
-  if (it != m_openContextMap.end()) 
+  tOpenContextMap::iterator it = m_openContextMap.find(exportName);
```
Comment (AlwinEsch) [inline src/NFSConnection.cpp]: Also here and somewhere else can be a `*::iterator` changed to a `auto`

File: src/NFSFile.cpp [hunk 7]
```diff
@@ -32,354 +30,235 @@ extern "C"
...
-    pItem.path = strdup(path.c_str());
-    pItem.label = strdup(currentExport.c_str());
-    pItem.title = NULL;
-
-    pItem.folder = true;
-    pItem.properties = NULL;
-    pItem.num_props = 0;
-    items.push_back(pItem);
-    ret = true; //added at least one entry
+    return NULL;
```
Comment (AlwinEsch) [inline src/NFSFile.cpp]: If you add a `NULL` can you direct use the new C++11 `nullptr`

File: src/NFSFile.cpp [hunk 8]
```diff
@@ -32,354 +30,235 @@ extern "C"
...
-
-//-- FreeSettings --------------------------------------------------------------
-// Free the settings struct passed from XBMC
-// !!! Add-on master function !!!
-//-----------------------------------------------------------------------------
-
-void ADDON_FreeSettings()
-{
-}
+  kodi::Log(ADDON_LOG_DEBUG,"CNFSFile::Open - opened %s", filename);
```
Comment (AlwinEsch) [inline src/NFSFile.cpp]: a `.c_str()` missing here.

File: src/NFSFile.cpp [hunk 9]
```diff
@@ -32,354 +30,235 @@ extern "C"
...
   result->exportPath = CNFSConnection::Get().GetContextMapId();
-  
-  ret = nfs_open(result->pNfsContext, filename.c_str(), O_RDONLY, &result->pFileHandle);
-  
-  if (ret != 0) 
+
+  if (bOverWrite)
   {
-    XBMC->Log(ADDON::LOG_INFO, "CNFSFile::Open: Unable to open file : '%s'  error : '%s'", url, nfs_get_error(result->pNfsContext));
+    kodi::Log(ADDON_LOG_INFO, "FileNFS::OpenForWrite() called with overwriting enabled! - %s", filename);
```
Comment (AlwinEsch) [inline src/NFSFile.cpp]: Here too.

**Comment by garbear** [comment]
@lrusak https://github.com/lrusak/vfs.nfs/pull/1

**Comment by lrusak** [comment]
made a bunch of changes. thanks all. I should probably squash soon.

Do all instances of `std::string` in `kodi::log` need to use `.c_str()`??

**Comment by garbear** [comment]
> Do all instances of std::string in kodi::log need to use .c_str()??

Yes, `kodi::log` uses printf-style `%s` which requires a char* and crashes if passed a `std::string`

**Comment by lrusak** [comment]
@garbear thanks, I'll fix it, although it doesn't look like it crashes it just produces garbage
```
13:20:13.340 T:140363499886336   DEBUG: AddOnLog: NFS support: NFS: Connected to server 192.168.1.106 and export ��^C��^?
```

**Comment by lrusak** [comment]
changed all relevant instances to use .c_str() (I hope)
```
13:48:27.894 T:140161922529024   DEBUG: AddOnLog: NFS support: NFS: Connected to server 192.168.1.106 and export /mnt/tank/concerts
```

**Comment by notspiff** [comment]
squash it up. others seem to have reviewed thoroughly.

**Comment by lrusak** [comment]
Squashed!

**Pull request merged**